src/System.Text.Primitives/System/Text/Encoding/Encoders.Utf8.cs

```diff
@@ -0,0 +1,481 @@
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    public static partial class Encoders
+    {
+        public static class Utf8
+        {
+            #region Tranforms
+
+            public static readonly Transformation FromUtf16 = new Utf16ToUtf8Transform();
+            public static readonly Transformation FromUtf32 = new Utf32ToUtf8Transform();
+
+            private sealed class Utf16ToUtf8Transform : Transformation
+            {
+                public override TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+                    => ConvertFromUtf16(source, destination, out bytesConsumed, out bytesWritten);
+            }
+
+            private sealed class Utf32ToUtf8Transform : Transformation
+            {
+                public override TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+                    => ConvertFromUtf32(source, destination, out bytesConsumed, out bytesWritten);
+            }
+
+            #endregion Transforms
+
+            #region Utf-16 to Utf-8 conversion
+
+            public static TransformationStatus ComputeEncodedBytesFromUtf16(ReadOnlySpan<byte> source, out int bytesNeeded)
+            {
+                bytesNeeded = 0;
+
+                // try? because Convert.ConvertToUtf32 can throw
+                // if the high/low surrogates aren't valid; no point
+                // running all the tests twice per code-point
+                try
+                {
+                    ref char utf16 = ref Unsafe.As<byte, char>(ref source.DangerousGetPinnableReference());
+                    int utf16Length = source.Length >> 1; // byte => char count
+
+                    for (int i = 0; i < utf16Length; i++)
+                    {
+                        var ch = Unsafe.Add(ref utf16, i);
+
+                        if ((ushort)ch <= 0x7f) // Fast path for ASCII
+                            bytesNeeded++;
+                        else if (!char.IsSurrogate(ch))
+                            bytesNeeded += EncodingHelper.GetUtf8EncodedBytes((uint)ch);
+                        else
+                        {
+                            if (++i >= utf16Length)
+                                return TransformationStatus.NeedMoreSourceData;
+
+                            uint codePoint = (uint)char.ConvertToUtf32(ch, Unsafe.Add(ref utf16, i));
+                            bytesNeeded += EncodingHelper.GetUtf8EncodedBytes(codePoint);
+                        }
+                    }
+
+                    if ((utf16Length << 1) != source.Length)
+                        return TransformationStatus.NeedMoreSourceData;
+
+                    return TransformationStatus.Done;
+                }
+                catch (ArgumentOutOfRangeException)
+                {
+                    return TransformationStatus.InvalidData;
```

Comment:
BTW, yesterday I was discussing some File APIs operating on spans and buffers with @stephentoub. It seems like we will want to use something like TranstromationStatus for such file i/o APIs. But file transfer is not really a transformation. Maybe we could come up with a better name for the enum. "OperationStatus" seems to generic. I would love a name that says "an operation on data in a buffer", but not a super long name.